core/sql/sort/Topn.cpp

```diff
@@ -0,0 +1,337 @@
+  if (stat)
+    bmoStats_ = stat->castToExBMOStats();
+  else
+    bmoStats_ = NULL;
+  if (bmoStats_)
+    bmoStats_->updateBMOHeapUsage((NAHeap *)heap_);
+}
+
+ 
+//------------------------------------------------------------------------
+// Class Destructor: Delete all the heap space pointed by pointers in Qsort
+//------------------------------------------------------------------------
+TopN::~TopN(void)
+{
+
+  for (int i = 0; i < runSize_; i++)
+	  topNKeys_[i].rec_->releaseTupp();
+
+  
+  if (topNKeys_ != NULL) {
+    NADELETEBASIC(topNKeys_, heap_);
+    topNKeys_ = NULL;
+  }
+  if (bmoStats_)
+    bmoStats_->updateBMOHeapUsage((NAHeap *)heap_);
+
+}
+
+Lng32 TopN::sortSend(void *rec, ULng32 len, void* tupp)		                        
+{
+  //if heap not built means, TopN array has more slots 
+  //available to fill. 
+  if(! isHeapified_)
+  {
+	  ex_assert(loopIndex_ >= 0, "TopN::sortSend: loopIndex_ is < 0");
+	  ex_assert(loopIndex_ < allocRunSize_, "TopN::sortSend: loopIndex_ > allocRunSize_");
+	  ex_assert(rec != NULL, "TopN::sortSend: rec is NULL");
+	
+	  Record * insertRec = new Record(rec, len, tupp, heap_, sortError_);
+	  
+	  topNKeys_[loopIndex_].key_ = insertRec->extractKey(keySize_, sortUtil_->config()->numberOfBytesForRecordSize());
+	  topNKeys_[loopIndex_].rec_ = insertRec;
+	  if (++loopIndex_  == allocRunSize_)
+	  {
+		 //Reaching here means, we have filled up the array. 
+		 //Now heapify the array to start accepting/eliminating new elements from now on.
+		 
+         //Note lookIndex_ contains the current number of filled elements.
+		  
+		 buildHeap();
+	  }
+	  return SORT_SUCCESS;
+  }
+  
+  //Reaching here means, heap is already build. 
+  //Check if the new rec belongs to this heap by comparing the
+  //new rec key with the root node of the heap ( root node is always the greatest).
+  insertRec(rec, len, tupp);
+  return SORT_SUCCESS;
+ }
+
+
+void TopN::buildHeap() 
+{
+	if(!isHeapified_)
+	{
+		//loopIndex_ is now <= TopN
+		runSize_ = loopIndex_;
+	
+		satisfyHeap();
+	
+		isHeapified_ = TRUE;
+	}
+}
+
+//Satisfy Heap makes sure the heap is balanced maxHeap.
+//Note this does not mean heap is sorted. It just makes sure
+//the higher level nodes are greater than lower level nodes.
+//Topmost node or root will be the highest.
+void TopN::satisfyHeap() 
+{
+	for (int i = (runSize_/2 ); i >= 0; i--)
+	    siftDown(topNKeys_, i, runSize_-1);
+}
+
+
+void TopN::insertRec(void *rec, ULng32 len, void* tupp) 
+{
+	ex_assert(isHeapified_, "TopN::insertRec: not heapified");
+
+	int root = 0; //Always, root node is the zero'th element in array.
+	
+	Record * insertRec = new Record(rec, len, tupp, heap_, sortError_);
+	insertRecKey_.key_ = insertRec->extractKey(keySize_, sortUtil_->config()->numberOfBytesForRecordSize());
+	insertRecKey_.rec_ = insertRec;
+
+	if (compare(topNKeys_[root].key_ ,insertRecKey_.key_) == KEY1_IS_GREATER)
+    {
+      swap( &topNKeys_[root],&insertRecKey_);
```

Comment:
It's places like this where the mixture of tabs and non-tabs makes the code hard to read due to odd indentation.